travis/apis.py

```diff
@@ -0,0 +1,181 @@
+        return response.json() if is_json else response
+
+
+class WeblateApi(Request):
+
+    def __init__(self):
+        super(WeblateApi, self).__init__()
+        self.repo_slug = None
+        self.branch = None
+        self._token = os.environ.get("WEBLATE_TOKEN")
+        self.host = os.environ.get(
+            "WEBLATE_HOST", "https://weblate.vauxoo.com/api")
+        self.tempdir = os.path.join(tempfile.gettempdir(), 'weblate_api')
+
+    def get_project(self, repo_slug, branch):
+        self.repo_slug = repo_slug
+        self.branch = branch
+        projects = self._request(self.host + '/projects')
+        for project in projects['results']:
+            if project['name'] == repo_slug:
+                return project
+        raise ApiException('No project found in "%s" for this path "%s"' %
+                           (self.host, repo_slug))
+
+    def load_project(self, repo_slug, branch):
+        self.project = self.get_project(repo_slug, branch)
+        self.load_components()
+
+    def get_components(self):
+        components = []
+        values = self._request(
+            self.host + '/projects/%s/components/' % self.project['slug'])
+        if not values['results']:
+            raise ApiException('No components found in the project "%s"' %
+                               self.project['slug'])
+        for value in values['results']:
+            if value['branch'] and value['branch'] != self.branch:
+                continue
+            components.append(value)
+        return components
+
+    def load_components(self):
+        self.components = self.get_components()
+
+    def pull(self):
+        pull = self._request(
+            self.host + '/projects/%s/repository/' % self.project['slug'],
+            {'operation': 'pull'})
+        return pull['result']
+
+    def component_repository(self, component, operation):
+        result = self._request(self.host + '/components/%s/%s/repository/' %
+                               (self.project['slug'], component['slug']),
+                               {'operation': operation})
+        return result['result']
+
+    @contextmanager
+    def components_context_manager(self):
+        try:
+            for component in self.components:
+                self.component_lock(component)
+            yield
+        finally:
+            for component in self.components:
+                self.component_unlock(component)
+
+    def component_lock(self, component):
+        while True:
+            url = (self.host + '/components/%s/%s/lock/' %
+                   (self.project['slug'], component['slug']))
+            lock = self._request(url)
+            if not lock['locked']:
+                lock = self._request(url, {'lock': True})
+                if lock['locked']:
+                    break
+            time.sleep(60)
+        return True
+
+    def component_unlock(self, component):
+        lock = self._request(self.host + '/components/%s/%s/lock/' %
```

Comment:
What about use the same `loop while` like as `component_lock`?
In fact, you could use the same method using a extra parameter to unlock/lock it